Devel core connector

dudoslav opened the pull request:
This PR adds implemntation of GRPC procedures. Some of the procedures are also tested.

dudoslav committed:
Implement ProtoHelper class

ProtoHelper class is used to convert GRPC structures
Appliance and Image to array of tags that are used by AWS.
The conversion can be done both ways.

dudoslav committed:
Implement FilterHelper class

FilterHelper class is used to generate filters that are used
by AWS to list specific resources. It simplifies use of AWS
describe_tags function.

dudoslav committed:
Implement CoreConnector class for GRPC protocol

CoreConnector class implements all procedures specified in
cloudkeeper.proto file.

amarthadan commented on .rubocop.yml:
```diff
@@ -1,15 +1,41 @@
+
+RSpec/InstanceVariable:
+  Exclude:
+    - 'spec/**/*.rb'
```
Where appropriate specify only the file violating the rule, not all the specs.

amarthadan commented on lib/cloudkeeper/aws/filter_helper.rb:
```diff
@@ -0,0 +1,39 @@
+module Cloudkeeper
+  module Aws
+    # Class used for generating filter structures used by AWS .describe_tags
+    class FilterHelper
```
Maybe try to refactor the class using the builder pattern so you can write something like `FilterHelper.only.by_name('name')`.

amarthadan replied:
Also these generic methods are great and you can use them but maybe write some more specific methods just for what you need since you're pretty much repeating 3 calls in `CoreConnector` anyway.

amarthadan commented on lib/cloudkeeper/aws/core_connector.rb:
```diff
@@ -0,0 +1,75 @@
+module Cloudkeeper
+  module Aws
+    # Class implementing GRPC procedures
+    class CoreConnector < CloudkeeperGrpc::Communicator::Service
```
I guess all the methods here need proper GRPC error handling.

amarthadan commented on lib/cloudkeeper/aws/proto_helper.rb:
```diff
@@ -0,0 +1,82 @@
+    # Module refining basic GRPC structs with additional methods
+    #   used for conversion from one format to another
+    module ProtoHelper
+      refine CloudkeeperGrpc::Appliance do
```
Maybe you can do some functional programming magic and make this more dynamically generated.

amarthadan commented on spec/acceptance/cloudkeeper/aws/grpc_api_spec.rb:
```diff
@@ -0,0 +1,172 @@
+        let(:image) { CloudkeeperGrpc::Image.new(uri: image_uri) }
+        let(:appliance) { CloudkeeperGrpc::Appliance.new(image: image) }
+
+        context 'with existing appliance' do
```
This is a bit strange naming.

dudoslav committed:
Modify .rubocop.yml to meet project standards

amarthadan commented on lib/cloudkeeper/aws/proto_helper.rb:
```diff
@@ -3,78 +3,52 @@ module Aws
-           { key: 'cloudkeeper_appliance_base_mpuri', value: base_mpuri },
-           { key: 'cloudkeeper_appliance_appid', value: appid },
-           { key: 'cloudkeeper_appliance_digest', value: digest }] + image.to_tags
+          tags = APPLIANCE_VALUES.map { |v| { key: "cloudkeeper_appliance_#{v}", value: send(v) } }
```
Remove code duplications and add `cloudkeeper_applaince` and `cloudkeeper_image` as constants.

dudoslav committed:
Add activesupport gem into project

This is required by cloudkeeper-grpc-ruby git submodule.

amarthadan commented on lib/cloudkeeper/aws/cloud.rb:
```diff
@@ -37,6 +37,15 @@ def upload_data(file_name, &block)
+          raise Cloudkeeper::Aws::Errors::Backend::BackendError,
+                "File #{file_name} does not exist"
+        end
+        obj.delete
```
Is this call blocking? Is it possible to know if it finished successfully?

dudoslav replied:
AWS error is raised if deletion fails. AWS error is then rescued and reraised as GRPCBadStatus.

amarthadan commented on lib/cloudkeeper/aws/core_connector.rb:
```diff
@@ -0,0 +1,90 @@
+          tag_descriptors = cloud.search_tags(FilterHelper.appliance(appliance.identifier))
+          unless tag_descriptors.size == 1
+            raise GRPC::BadStatus.new(CloudkeeperGrpc::Constants::STATUS_CODE_INVALID_RESOURCE_STATE,
+                                      'Appliance duplication or not found')
```
Try a bit more descriptive error message.

amarthadan commented on lib/cloudkeeper/aws/core_connector.rb:
```diff
@@ -0,0 +1,90 @@
+module Cloudkeeper
+  module Aws
+    # Class implementing GRPC procedures
+    class CoreConnector < CloudkeeperGrpc::Communicator::Service
```
Please rework appliance update into two separate methods according to new proto file (sorry about that).

dudoslav replied:
No need to be sorry, I'll look into that

dudoslav committed:
Change core_connector to resemble proto file

Specification cloudkeeper.proto was changes so core_connector has to
implement additional procedure.

amarthadan reviewed (approved):
:ramen:
